vispy/visuals/visual.py

```diff
@@ -3,52 +3,164 @@
 # Distributed under the (new) BSD License. See LICENSE.txt for more info.
 
 from __future__ import division
+import weakref
 
 from ..util.event import EmitterGroup, Event
-from .shaders import StatementList
+from .shaders import StatementList, MultiProgram
+from .transforms import TransformSystem
 from .. import gloo
 
-"""
-API Issues to work out:
 
-  * Need Visual.bounds() as described here:
-    https://github.com/vispy/vispy/issues/141
-
-"""
-
-
-class Visual(object):
+class VisualShare(object):
+    """Contains data that is shared between all views of a visual.
+    
+    This includes:
+    
+    * GL state variables (blending, depth test, etc.)
+    * A weak dictionary of all views
+    * A list of filters that should be applied to all views
+    * A cache for bounds.
     """
-    Abstract class representing a drawable object.
-
-    At a minimum, Visual subclasses should extend the draw() method. 
-
-    Events:
+    def __init__(self):
+        # Note: in some cases we will need to compute bounds independently for each
+        # view. That will have to be worked out later..
+        self.bounds = {}
+        self.gl_state = {}
+        self.views = weakref.WeakValueDictionary()
+        self.filters = []
 
-    update : Event
-        Emitted when the visual has changed and needs to be redrawn.
-    bounds_change : Event
-        Emitted when the bounds of the visual have changed.
 
+class BaseVisual(object):
+    """Superclass for all visuals.
+    
+    This class provides:
+    
+    * A TransformSystem.
+    * Two events: `update` and `bounds_change`.
+    * Minimal framework for creating views of the visual.
+    * A data structure that is shared between all views of the visual.
+    * Abstract `draw`, `bounds`, `attach`, and `detach` methods.
+    
     """
-
-    def __init__(self):
-        self._visible = True
+    def __init__(self, vshare=None, key=None):
+        self._view_class = getattr(self, '_view_class', VisualView)
+        
+        if vshare is None:
+            vshare = VisualShare()
+            assert key is None
+            key = 'default'
+        
+        self._vshare = vshare
+        self._view_key = key
+        self._vshare.views[key] = self
+        
         self.events = EmitterGroup(source=self,
                                    auto_connect=True,
                                    update=Event,
                                    bounds_change=Event,
                                    )
-        self._gl_state = {'preset': None}
-        self._filters = set()
+        
+        self.transforms = TransformSystem()
+
+    @property
+    def transform(self):
+        return self.transforms.visual_to_document
+    
+    @transform.setter
+    def transform(self, tr):
+        self.transforms.visual_transform = tr
+
+    def view(self, key=None):
+        """Return a new view of this visual.
+        """
+        if key is None:
```

Comment:
The keys were just intended as unique names for debugging purposes. They ended up being useless so I have removed them :)